[postfilter-impl-4] Move Preempt() to defaultpreemption package.

Huang-Wei opened the pull request:
**What type of PR is this?**

/kind cleanup
/sig scheduling

**What this PR does / why we need it**:

This PR moves Preempt() and its related functions from pkg/scheduler/core to defaultpreemption package.

**Which issue(s) this PR fixes**:

Part of #91038

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

ahg-g commented on pkg/scheduler/scheduler.go:
```diff
@@ -480,19 +480,19 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
+
+				// Run PostFilter plugins to try to make the pod schedulable in a future scheduling cycle.
+				result, status := prof.RunPostFilterPlugins(ctx, state, pod, fitError.FilteredNodesStatuses)
+
 				metrics.PreemptionAttempts.Inc()
 				metrics.SchedulingAlgorithmPreemptionEvaluationDuration.Observe(metrics.SinceInSeconds(preemptionStartTime))
 				metrics.DeprecatedSchedulingDuration.WithLabelValues(metrics.PreemptionEvaluation).Observe(metrics.SinceInSeconds(preemptionStartTime))
```
those should move to the preemption plugin

Huang-Wei replied:
This is now covered in the previous PR ([postfilter-impl-3]).

ahg-g commented on pkg/scheduler/scheduler.go:
```diff
@@ -480,19 +480,19 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
 		nominatedNode := ""
 		if fitError, ok := err.(*core.FitError); ok {
-			if sched.DisablePreemption {
+			if sched.DisablePreemption || !prof.HasPostFilterPlugins() {
 				klog.V(3).Infof("Pod priority feature is not enabled or preemption is disabled by scheduler configuration." +
 					" No preemption is performed.")
 			} else {
 				preemptionStartTime := time.Now()
```
we should move this inside the preemption plugin in a follow up PR, this else clause should just RunPostFilterPlugins.

ahg-g commented on pkg/scheduler/scheduler.go:
```diff
@@ -480,19 +480,19 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
 		nominatedNode := ""
 		if fitError, ok := err.(*core.FitError); ok {
-			if sched.DisablePreemption {
+			if sched.DisablePreemption || !prof.HasPostFilterPlugins() {
```
In a followup task, we should deprecate ```DisablePreemption``` since preemption can now be disabled by disabling the default preemption plugin itself.

Huang-Wei replied:
Yes, I listed it in the umbrella issue already.

ahg-g commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -50,13 +62,530 @@ func New(_ runtime.Object, fh framework.FrameworkHandle) (framework.Plugin, erro
+// other pods with the same priority. The nominated pod prevents other pods from
+// using the nominated resources and the nominated pod could take a long time
+// before it is retried after many other pending pods.
+func Preempt(ctx context.Context, fh framework.FrameworkHandle, state *framework.CycleState, pod *v1.Pod, m framework.NodeToStatusMap) (string, error) {
```
don't export?

Huang-Wei replied:
I think it depends on which kind of developers we care about:

- If the developers are following the scheduler framework, they don't need to call this function - as in that case, they just use the defaultpreemption plugin; otherwise, they built their "preempt" function using combined built-in preemption libraries and customized ones. 
- The other case is free-style scheduler implementation, they may want to call it directly as they're not using scheduler framework.

I'll evaluate whether this function, as well as its internal functions, needs to be made public, in the next PR.

ahg-g replied:
sounds good, until then I think we should just follow the normal golang best practices which tells us that this should be internal.

Huang-Wei replied:
Yes, making it private makes more sense. Let me change it in this PR later in case it's out of track.

ahg-g commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -50,13 +62,530 @@ func New(_ runtime.Object, fh framework.FrameworkHandle) (framework.Plugin, erro
+	return &framework.PostFilterResult{NominatedNodeName: nnn}, framework.NewStatus(framework.Success)
+}
+
+// Preempt finds nodes with pods that can be preempted to make room for "pod" to
```
I am guessing all this logic was simply moved, nothing at all changed, right?

Huang-Wei replied:
Yes, just code movement. A dependent PR (#92107) also just moves testing stuff to pkg/scheduler/testing.

Huang-Wei commented:
/hold 
for dependent PRs to get merged first.

Huang-Wei committed:
Move Preempt() and its related functions to defaultpreemption package

Refactor genericScheduler and signature of preemption funcs
  - remove podNominator from genericScheduler
  - simplify signature of preemption functions

Make Preempt() private

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

ahg-g commented:
Parent PRs merged, can you please rebase.

Huang-Wei commented:
/retest

ahg-g commented on pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -585,10 +415,9 @@ func addNominatedPods(ctx context.Context, pr framework.PluginsRunner, nominator
 // SelectVictimsOnNode(). Preempt removes victims from PreFilter state and
 // NodeInfo before calling this function.
-func podPassesFiltersOnNode(
+func PodPassesFiltersOnNode(
```
what is the plan for this function? I don't think we want to keep the dependency form the plugin to generic_scheduler.go.

Huang-Wei replied:
In addition to the preemption path, this function is called by the regular Filter path (`findNodesThatPassFilters()`). I think the destination should be the runtime package (pkg/scheduler/framework/runtime).

If we do so, the last dependency is `core.ErrNoNodesAvailable`, which is easier to sort out.

ahg-g replied:
hmm, we don't want plugins to have dependencies on the runtime pkg either. We should probably have an interface to invoke it to avoid that dependency.

Huang-Wei replied:
Let me put a TODO here for further brainstorming.

ahg-g commented on pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -550,12 +380,12 @@ func (g *genericScheduler) findNodesThatPassExtenders(pod *v1.Pod, filtered []*v
 // 3) augmented nodeInfo.
-func addNominatedPods(ctx context.Context, pr framework.PluginsRunner, nominator framework.PodNominator, pod *v1.Pod, state *framework.CycleState, nodeInfo *framework.NodeInfo) (bool, *framework.CycleState, *framework.NodeInfo, error) {
-	if nominator == nil || nodeInfo == nil || nodeInfo.Node() == nil {
+func addNominatedPods(ctx context.Context, ph framework.PreemptHandle, pod *v1.Pod, state *framework.CycleState, nodeInfo *framework.NodeInfo) (bool, *framework.CycleState, *framework.NodeInfo, error) {
```
ditto, what is the plan for this function?

Huang-Wei replied:
Per https://github.com/kubernetes/kubernetes/pull/92108/files#r444659410, it should be moved along with `PodPassesFiltersOnNode()`, and remain private.

ahg-g reviewed (commented):
nits

ahg-g commented:
Thanks Wei, please squash.

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
Ah, seems Aldo's new PR records latency inside each filter plugin, so the refactoring in this PR to change `profile` to `framework` is failing. Will take a look tomorrow.

alculquicondor commented:
> Ah, seems Aldo's new PR records latency inside each filter plugin, so the refactoring in this PR to change `profile` to `framework` is failing. Will take a look tomorrow.

Per-profile latencies are just recorded in the framework (not the plugins). Note that I added the profile name to the framework implementation. I think it makes sense to merge framework and profile (ie, get rid of the `Profile` type), given that framework holds all the profile needs. I made sense while the framework package was growing, but not so much anymore now that it was split into the runtime package.

But let's do that cleanup in 1.20.

Huang-Wei commented:
> I think it makes sense to merge framework and profile (ie, get rid of the Profile type), given that framework holds all the profile needs.

Exactly.

> Per-profile latencies are just recorded in the framework (not the plugins).

True, I should have said "... Aldo's new PR records latency upon calling filter plugins ..." :)

Huang-Wei commented:
@ahg-g Rebased and fell back to keep prof as the parameter of some functions.

The pure diff (comparing to your last review) can be viewed here: https://gist.github.com/Huang-Wei/61ff756cb5a3cb4bfe63026acada1855.

ahg-g commented:
/lgtm

Congrats!

Huang-Wei commented:
/hold cancel

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest
